src/v/cloud_topics/level_one/compaction/source.cc

```diff
@@ -122,103 +176,114 @@ ss::future<ss::stop_iteration> compaction_source::map_building_iteration() {
+        map_is_full = res.map_is_full;
+        auto max_indexed_offset = res.max_indexed_offset;
+
+        if (max_indexed_offset.has_value()) {
+            bool range_has_tombstones = res.range_has_tombstones;
+            auto base_offset = start_offset;
+            auto last_offset = map_is_full ? model::offset_cast(
+                                               max_indexed_offset.value())
+                                           : max_offset;
+            dassert(
+              base_offset <= last_offset,
+              "Cleaned range must be properly bounded.");
+            _new_cleaned_ranges.push_back(
+              {.base_offset = base_offset,
+               .last_offset = last_offset,
+               .has_tombstones = range_has_tombstones});
+        }
+
+        if (map_is_full) {
+            break;
+        }
     }
 
-    if (min_offset_indexed.has_value() && max_offset_indexed.has_value()) {
-        _indexed_intervals.insert(
-          min_offset_indexed.value(), max_offset_indexed.value());
+    if (map_is_full) {
+        co_return ss::stop_iteration::yes;
     }
 
-    ++_map_building_it;
-    co_return ret;
+    ++_dirty_range_it;
+    co_return ss::stop_iteration::no;
 }
 
 ss::future<ss::stop_iteration> compaction_source::deduplication_iteration(
   compaction::sliding_window_reducer::sink& sink) {
+    auto& ct_sink = static_cast<compaction_sink&>(sink);
     if (preempted()) {
         co_return ss::stop_iteration::yes;
     }
 
-    if (_indexed_intervals.empty()) {
-        // We indexed nothing.
+    if (_extents_it == _extents_end_it) {
         co_return ss::stop_iteration::yes;
     }
 
-    // TODO: Find a way to chunk this work into iterations.
-    // An easy way to do this is to go from [start_offset,
-    // next_dirty_interval.last_offset], and so on.
-    auto indexed_intervals = _indexed_intervals.to_vec();
-    kafka::offset start_offset{0};
-    kafka::offset max_offset{indexed_intervals.back().last_offset};
-    cloud_topic_log_reader_config config(start_offset, max_offset, _as);
-    auto rdr = model::record_batch_reader(
-      std::make_unique<level_one_log_reader_impl>(
-        config, _ntp, _tp, _metastore, _io));
-
-    auto stats = co_await rdr.consume(
-      compaction_filter{sink, *_map, _ntp, _removable_tombstone_ranges},
-      model::no_timeout);
-    if (stats.has_removed_data()) {
-        vlog(
-          compaction_log.info,
-          "L1 compaction removing data from CTP {}, offset range ({}~{}), "
-          "stats: {}",
-          start_offset,
-          max_offset,
-          _ntp,
-          stats);
-    } else {
-        vlog(
-          compaction_log.debug,
-          "L1 compaction not removing data from CTP {}, offset range ({}~{}), "
-          "stats: {}",
-          start_offset,
-          max_offset,
-          _ntp,
-          stats);
+    auto& extent = *_extents_it;
+    if (should_compact_extent(extent, _min_compaction_lag_ms)) {
```

Comment:
Should we stop iterating if we get to an extent that shouldn't be compacted?

Reply:
Ah fair thanks, I momentarily forgot how the metastore compaction metadata worked :)